src/compute/src/logging/timely.rs

```diff
@@ -526,3 +325,406 @@ fn create_address_row(id: usize, worker: WorkerIdentifier, address: &[usize]) ->
+
+                let datum = ScheduleHistogramDatum {
+                    operator: event.id,
+                    duration_pow: 1 << bucket,
+                };
+                let diff = Diff::cast_from(-count);
+                self.output
+                    .schedules_histogram
+                    .give(self.cap, (datum, ts, diff));
+            }
+        }
+
+        if operator.addr.len() == 1 {
+            let dataflow_id = operator.addr[0];
+            self.handle_dataflow_shutdown(dataflow_id);
+        }
+    }
+
+    fn handle_dataflow_shutdown(&mut self, dataflow_id: usize) {
+        // When a dataflow shuts down, we need to retract all its channels.
+
+        let Some(channels) = self.state.dataflow_channels.remove(&dataflow_id) else {
+            return;
+        };
+
+        let ts = self.ts();
+        for channel in channels {
+            // Retract channel description.
+            let datum = ChannelDatum {
+                id: channel.id,
+                source: channel.source,
+                target: channel.target,
+            };
+            self.output.channels.give(self.cap, (datum, ts, -1));
+
+            let datum = AddressDatum {
+                id: channel.id,
+                address: channel.scope_addr,
+            };
+            self.output.addresses.give(self.cap, (datum, ts, -1));
+
+            // Retract messages logged for this channel.
+            if let Some(sent) = self.state.messages_sent.remove(&channel.id) {
+                for (target_worker, count) in sent.iter().enumerate() {
+                    let datum = MessageDatum {
+                        channel: channel.id,
+                        worker: target_worker,
+                    };
+                    self.output
+                        .messages_sent
+                        .give(self.cap, (datum, ts, -count));
+                }
+            }
+            if let Some(received) = self.state.messages_received.remove(&channel.id) {
+                for (source_worker, count) in received.iter().enumerate() {
+                    let datum = MessageDatum {
+                        channel: channel.id,
+                        worker: source_worker,
+                    };
+                    self.output
+                        .messages_received
+                        .give(self.cap, (datum, ts, -count));
+                }
+            }
+        }
+    }
+
+    fn handle_park(&mut self, event: ParkEvent) {
+        let time_ns = self.time.as_nanos();
+        match event {
+            ParkEvent::Park(requested) => {
+                let park = Park { time_ns, requested };
+                let existing = self.state.last_park.replace(park);
+                if existing.is_some() {
+                    error!("park without a succeeding unpark");
+                }
+            }
+            ParkEvent::Unpark => {
+                if let Some(park) = self.state.last_park.take() {
```

Comment:
Nit: Conversely, this would be a good candidate for a `let ... else` construct, but not sure if a `return` here would break only out of the match arm's block expression?

Reply:
No, a `return` returns from the function, but that's still fine in this case.